### PR TITLE
chore: test tracker validation service TECH-880

### PR DESCRIPTION
**Pull request by teleivo**
* make tests easily discover able from IDE. Prefixing them with DefaultTrackerValidationService...
* merge the 2 unit test classes into one DefaultTrackerValidationServiceTest
* add tests that test the different "configurations" an AbstractTrackerDtoValidationHook hook can have: removeOnError, needsToRun
* add test that shows the relationships between entities like TEI -> enrollments -> events do not have an influence on hook execution. Meaning children of an invalid TEI still get validated.
* extract helpers to assert on errors/warnings in TrackerValidationReport like we have them for the ValidationErrorReporter

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/hooks/AssertTrackerValidationReport.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+package org.hisp.dhis.tracker.validation.hooks;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import org.hisp.dhis.tracker.TrackerType;
+import org.hisp.dhis.tracker.domain.TrackerDto;
+import org.hisp.dhis.tracker.report.TrackerErrorCode;
+import org.hisp.dhis.tracker.report.TrackerValidationReport;
+
+public class AssertTrackerValidationReport
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/hooks/AssertTrackerValidationReport.java:37]: Is any of the code here test specific? I wonder why not move the methods from static methods here to become member methods of `TrackerValidationReport`. This is good way to have people also find these methods elsewhere any these might be useful even in non-test code.

**Review by jbee — COMMENTED** [review]
Why combine two test classes when they have different setup? I don't see how putting it in one file improves readability. We have various classes where we have multiple test classes testing different aspects of it which generally helps when there are many tests.

**Comment by teleivo** [comment]
> Why combine two test classes when they have different setup? I don't see how putting it in one file improves readability. We have various classes where we have multiple test classes testing different aspects of it which generally helps when there are many tests.

I don't think the 3 tests using mocks only warrant their own test class. That's why I moved them into the same class as all the tests are unit tests, some using only mocks, some using one mock per test. Two test classes also require me to come up with 2 names for them for which I have no good answer, which again tells me => one test class. But I don't feel super strongly about this and can split them again if that's preferred by someone else as well.

File: dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/hooks/AssertTrackerValidationReport.java (hunk 1 above)
Comment (teleivo) [reply dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/hooks/AssertTrackerValidationReport.java:37]: Yes, this code is only test-specific. It wraps a few assertions into one method using `hasError(Predicate)` which is already a member of the TrackerValidationReport.

Comment (jbee) [reply dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/hooks/AssertTrackerValidationReport.java:37]: Right, totally missed that you were making assertions in the methods. 
I think we should stick to the pattern and name asserting methods in a way that is fully explanatory when imported with a static import. Something like `assertHasError` and so forth.

**Review by jbee — APPROVED** [review]

**Review by stian-sandvold — APPROVED** [review]

**Pull request merged**